### PR TITLE
Fix #9025: turn non-distinct single columns into count(*), throw for …

### LINKED ISSUE #9025: Panache Paging does not support projections
**Describe the bug**
When using projections or retrieve specific fields from a query using pagination the generated `count` query captures incorrectly the fields / projection. Hence, producing a malformed query that throws a hibernate query.

**Expected behavior**
Either capture correctly the fields / projection or use the `count(*)` for the count query.

**Actual behavior**
`org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: org.hibernate.hql.internal.ast.QuerySyntaxException: expecting CLOSE, found ',' near line 1` exception is throw.

**To Reproduce**
1. Create query in the form "select a.id, b.field1 from A a joing B b on a.id = b.id"
2. Configure .page(1,25)
3. Try to do pageCount()

**Configuration**
```properties
# Add your application.properties here, if applicable.

```

**Screenshots**

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Darwin hguerrer-osx 18.7.0 Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64 x86_64
 - Output of `java -version`: 
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment GraalVM CE 19.3.1 (build 11.0.6+9-jvmci-19.3-b07)
OpenJDK 64-Bit Server VM GraalVM CE 19.3.1 (build 11.0.6+9-jvmci-19.3-b07, mixed mode, sharing)
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 
1.4.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /Users/hguerrer/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /Users/hguerrer/.sdkman/candidates/java/19.3.1.r11-grl
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.14.6", arch: "x86_64", family: "mac"

**Additional context**
(Add any other context about the problem here.)
Using panache with projections

### PR DESCRIPTION
**Pull request by FroMage**
…distinct multi columns

This will support:

- non-distinct single/multi-column selects by turning them into count(*)
- distinct single column by turning them into count(distinct col)
- from.* into count(*)...

And barf at unsupported use-cases such as select distinct col1, col2...

### CODE REVIEW

**Comment by FroMage** [comment]
I've asked @Sanne for review, but don't be too pedantic over support for counting here, ok? I'm trying to support the easy/common use-cases and throw on more complex ones until someone requires supporting them :)

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java [hunk 1]
```diff
@@ -198,9 +204,24 @@ private String countQuery() {
...
-            countQuery = "SELECT COUNT(" + selectMatcher.group(1) + ") " + selectMatcher.group(2);
-        } else {
+            // this one cannot be null
+            String firstSelection = selectMatcher.group(1).trim();
+            if (firstSelection.toLowerCase().startsWith("distinct ")) {
+                // this one can be null
+                String secondSelection = selectMatcher.group(2);
+                // we can only count distinct single columns
+                if (secondSelection != null && !secondSelection.trim().isEmpty()) {
+                    throw new IllegalArgumentException("Count query not supported for select query: " + query);
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:214]: Maybe throw a `PanacheQueryException` instead ?

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java [hunk 2]
```diff
@@ -198,9 +204,24 @@ private String countQuery() {
...
+                }
+                countQuery = "SELECT COUNT(" + firstSelection + ") " + selectMatcher.group(3);
+            } else {
+                // it's not distinct, forget the column list
+                countQuery = "SELECT COUNT(*) " + selectMatcher.group(3);
+            }
+        } else if (FROM_PATTERN.matcher(query).matches()) {
             countQuery = "SELECT COUNT(*) " + query;
+        } else {
+            throw new IllegalArgumentException("Count query not supported for select query: " + query);
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:224]: Maybe throw a `PanacheQueryException` instead ?

**Review by loicmathieu — APPROVED** [review]
LGTM, I made some suggestion for the exception thrown but I'm OK anyway.

I think we would need a way for the user to give its count query in order to be able to handle complex use cases ...

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java (hunk 2 above)
Comment (FroMage) [reply extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java]: Fine, done.

**Comment by FroMage** [comment]
> I think we would need a way for the user to give its count query in order to be able to handle complex use cases ...

Let's talk about this when someone requests this.

**Comment by loicmathieu** [comment]
@FroMage nothing more for me ;)

**Comment by FroMage** [comment]
@loicmathieu can you re-accept your review? For some reason it's not counted anymore.

**Comment by loicmathieu** [comment]
@FroMage I don'y have write access, you need one reviewer with write access ...

**Comment by FroMage** [comment]
Oh that's why. Thanks.

@geoand will you be my reviewer please? :)

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java [hunk 3]
```diff
@@ -198,9 +204,24 @@ private String countQuery() {
...
+                }
+                countQuery = "SELECT COUNT(" + firstSelection + ") " + selectMatcher.group(3);
+            } else {
+                // it's not distinct, forget the column list
+                countQuery = "SELECT COUNT(*) " + selectMatcher.group(3);
+            }
+        } else if (FROM_PATTERN.matcher(query).matches()) {
             countQuery = "SELECT COUNT(*) " + query;
+        } else {
+            throw new PanacheQueryException("Count query not supported for select query: " + query);
```
Comment (geoand) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:224]: I generally think it's a good idea to enclose this type of output in quotes, but I'll leave it up to you :)

**Review by geoand — APPROVED** [review]
:+1:

**Comment by FroMage** [comment]
Thanks man.

**Pull request merged**